biodatasets/osiris/osiris.py

```diff
@@ -0,0 +1,302 @@
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager) -> List[datasets.SplitGenerator]:
+
+        urls = _URLS[_DATASETNAME]
+        data_dir = dl_manager.download(urls)
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                # Whatever you put in gen_kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir[0]),
+                    "split": "data",
+                },
+            )
+        ]
+
+    def _get_offsets(self, parent: ET.Element, child: ET.Element) -> List[int32]:
+        '''
+        Retrieves character offsets for child from parent.
+        '''
+        parent_text = " ".join([" ".join([t for t in c.itertext()])
+                               for c in list(parent) if c.tag != "Pmid"])
+        child_text = " ".join([t for t in child.itertext()])
+        start = parent_text.index(child_text)
+        end = start + len(child_text)
+        return [start, end]
+
+    def _get_dict(self, elem: ET.Element) -> dict:
+        '''
+        Retrieves dict from XML element.
+        '''
+        elem_d = dict()
+        for child in elem:
+            elem_d[child.tag] = {}
+            elem_d[child.tag]["text"] = " ".join([t for t in child.itertext()])
+
+            if child.tag != "Pmid":
+                elem_d[child.tag]["offsets"] = self._get_offsets(elem, child)
+
+            for c in child:
+                elem_d[c.tag] = []
+
+            for c in child:
+                c_dict = c.attrib
+                c_dict["offsets"] = self._get_offsets(child, c)
+                elem_d[c.tag].append(c.attrib)
+
+        return elem_d
+
+    def _handle_missing_variants(self, row: dict) -> dict:
+        '''
+        If variant is not present in the row this function adds one variant 
+        with no data (to make looping though items possible) and returns the new row.
+        Otherwise returns unchanged row.
+        '''
+
+        if row.get("variant", 0) == 0:
+            row["variant"] = [
+                {
+                    "v_id": "",
+                    "v_lex": "",
+                    "offsets": [0, 0]
+                }
+            ]
+        return row
+
+    def _get_entities(self, row: dict) -> List[dict]:
+        '''
+        Retrieves two lists of dicts for genes and variants. 
+        After that, chains both together.
+        '''
+        genes = [
+            {
+                "id": str(uuid.uuid4()),
+                "offsets": [gene["offsets"]],
+                "text": [" ".join([gene["g_id"], gene["g_lex"]])],
+                "type": "gene",
+                "normalized": []
+            } for gene in row["gene"]
+        ]
+        variants = [
+            {
+                "id": str(uuid.uuid4()),
+                "offsets": [variant["offsets"]],
```

Comment:
Similar as above, although a bit trickier.
Variants that have no dnSNP are "normalized" with a HGVS-like notation. E.g.
```html
<variant v_id="No" v_lex="+1632 T" v_norm="1632T">
```
In this case the entity should be: 
```python
{
"text" : gene["v_lex"],
"normalized" : [{"db_name" : "HGVS-like", "db_id" : variant["v_norm"]}]
}
```
Variants for which there is a dbSNP, e.g. 
```html
<variant v_id="660339" v_lex="Ala55Val" v_norm="A55V">Ala55Val</variant>
``` 
The entity should be:
```python
{
"text" : gene["v_lex"],
"normalized" : [{"db_name" : "dbSNP", "db_id" : variant["v_id"]}]
}
```
and we safely loose the HGVS-like notation, since it is recoverable from the dbSNP id. In the `source` schema we keep both to preserve the original dataset

Reply:
Changed my code to reflect variants and genes correctly.